# 6.1 ANSWERS

1.  IMPORT LIBRARIES AND DATA SETS

2. CHECKING DATA SOURCES ( SHAPE, HEAD)

3. CLEANING DATA ( MISSING VALUES AND DUPLICATES)

4. UNDERSTANDING YOUR DATA 

# 1.  IMPORT LIBRARIES AND DATA SETS

In [3]:
# import all libraries

import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

In [4]:
# creating a shortcut to access the Airbnb data as path

path= r'C:\Users\carol\Documents\Airbnb'

In [6]:
# importing original data set for listings 

df_listings = pd.read_csv(os.path.join(path, 'Original Data', 'listings.csv.gz'))

In [10]:
# importing original data set for listings (short version)

df_listing_st= pd.read_csv(os.path.join(path, 'Original Data', 'listings.csv'))

In [7]:
# importing original data set for reviews

df_reviews= pd.read_csv(os.path.join(path, 'Original Data', 'reviews.csv.gz'))


# 2. CHECKING DATA SOURCES ( SHAPE, HEAD)

In [15]:
df_listings.shape

(26924, 75)

In [9]:
#Checking df_listings columns

df_listings.head(5)


id                                       listing_url  \
0  1104967621421098322  https://www.airbnb.com/rooms/1104967621421098322   
1   930941073152690361   https://www.airbnb.com/rooms/930941073152690361   
2             51647382             https://www.airbnb.com/rooms/51647382   
3  1026072412626070839  https://www.airbnb.com/rooms/1026072412626070839   
4   766423030884644417   https://www.airbnb.com/rooms/766423030884644417   

        scrape_id last_scraped       source                              name  \
0  20240615011404   2024-06-15  city scrape  Habitación de lujo, baño privado   
1  20240615011404   2024-06-15  city scrape            habitación para pareja   
2  20240615011404   2024-06-15  city scrape            Buenos Aires en Madrid   
3  20240615011404   2024-06-15  city scrape            Habitación Madrid Rio!   
4  20240615011404   2024-06-15  city scrape           habitación para turista   

                                         description  neighborhood_overview  \
0                                                NaN                    NaN   
1  Rent room for loose nights or weekends  to a c...                    NaN   
2  This special place is close to everything, mak...  The neighborhood is m   
3  Enjoy the simplicity of this quiet and central...                    NaN   
4  Enjoy a stylish experience at this centrally-l...                    NaN   

                                         picture_url    host_id  ...  \
0  https://a0.muscache.com/pictures/miso/Hosting-...  565307927  ...   
1  https://a0.muscache.com/pictures/85f0b383-e8f8...  442944608  ...   
2  https://a0.muscache.com/pictures/db018af4-f622...  252336445  ...   
3  https://a0.muscache.com/pictures/miso/Hosting-...  545435841  ...   
4  https://a0.muscache.com/pictures/miso/Hosting-...    3882456  ...   

  review_scores_communication review_scores_location review_scores_value  \
0                        5.00                   5.00                5.00   
1                        5.00                   4.67                5.00   
2                        4.83                   4.51                4.61   
3                        5.00                   5.00                5.00   
4                         NaN                    NaN                 NaN   

  license instant_bookable calculated_host_listings_count  \
0     NaN                f                              1   
1     NaN                f                              2   
2     NaN                t                              2   
3     NaN                f                              1   
4     NaN                f                              1   

  calculated_host_listings_count_entire_homes  \
0                                           0   
1                                           0   
2                                           0   
3                                           0   
4                                           0   

  calculated_host_listings_count_private_rooms  \
0                                            1   
1                                            2   
2                                            2   
3                                            1   
4                                            1   

  calculated_host_listings_count_shared_rooms reviews_per_month  
0                                           0              0.61  
1                                           0              0.28  
2                                           0              5.40  
3                                           0              0.14  
4                                           0               NaN  

[5 rows x 75 columns]

In [17]:
df_listing_st.shape

(26924, 18)

In [14]:
#Checking df_listings_st data 

df_listing_st.head()

id                                               name    host_id  \
0  21853                               Bright and airy room      83531   
1  24805                             Gran Via Studio Madrid  346366726   
2  40718  COZY APARTMENT CIBELES-GRAN VIA-PTA DEL SOL-CH...   34217435   
3  40916                    Holiday Apartment Madrid Center     130907   
4  30320                        Great Vacational Apartments     130907   

  host_name neighbourhood_group neighbourhood  latitude  longitude  \
0     Abdel              Latina      Cármenes  40.40381   -3.74130   
1    Alvaro              Centro   Universidad  40.42183   -3.70529   
2      Luis              Centro      Justicia  40.41969   -3.69736   
3      Dana              Centro   Universidad  40.42247   -3.70577   
4      Dana              Centro           Sol  40.41476   -3.70418   

         room_type  price  minimum_nights  number_of_reviews last_review  \
0     Private room   31.0               4                 33  2018-07-15   
1  Entire home/apt  104.0               5                 43  2024-06-01   
2  Entire home/apt  114.0               2                150  2024-05-30   
3  Entire home/apt    NaN               5                 49  2021-12-11   
4  Entire home/apt    NaN               5                172  2022-09-26   

   reviews_per_month  calculated_host_listings_count  availability_365  \
0               0.28                               2                83   
1               0.37                               1               113   
2               0.89                              15               310   
3               0.30                               3                 0   
4               1.01                               3                 0   

   number_of_reviews_ltm license  
0                      0     NaN  
1                     13     NaN  
2                     20     NaN  
3                      0     NaN  
4                      0     NaN

In [18]:
df_reviews.shape

(1189990, 6)

In [13]:
#Checking df_reviews

df_reviews.head()

listing_id        id        date  reviewer_id reviewer_name  \
0       21853  21051116  2014-10-10      8506071    Pedro Abel   
1       21853  21268157  2014-10-13     19062863          Olga   
2      576487   2576669  2012-10-10      3732963         Fabio   
3      576487   2954302  2012-11-24      3892202            Mo   
4       21853  22527396  2014-11-09      6099461          Hugh   

                                            comments  
0  Mi experiencia en casa de Adel fue buena, aunq...  
1   Adel is a very warm person and thoughtful! He...  
2  Pilar è stata gentile e molto disponibile. Abb...  
3  Merci Pilar. Ton appartement est très bien, pr...  
4  Adel is an exceptionally welcoming host. I was...

# 3. CLEANING DATA ( MISSING VALUES AND DUPLICATES)

Conduct some basic data cleaning and consistency checks in Jupyter to ensure your data is ready for further analysis.

In [21]:
# Checking for any missing the df_listing_st

missing_values=df_listing_st.isnull().sum()
print("Missing Values in Each Column:\n", missing_values)


Missing Values in Each Column:
 id                                    0
name                                  0
host_id                               0
host_name                             4
neighbourhood_group                   0
neighbourhood                         0
latitude                              0
longitude                             0
room_type                             0
price                              5304
minimum_nights                        0
number_of_reviews                     0
last_review                        5856
reviews_per_month                  5856
calculated_host_listings_count        0
availability_365                      0
number_of_reviews_ltm                 0
license                           24182
dtype: int64


In [22]:
# Checking for duplicates df_listing_st
duplicate_rows = df_listing_st.duplicated().sum()
print("Number of Duplicate Rows:", duplicate_rows)

Number of Duplicate Rows: 0


**NOTE**: *the df_listing_st shows 5304 price values edited as NAN and 24,182 listings with no License.
A tourist licence is required in Spain allowing you to rent out the property as a tourist accomodation. However Airbnb is a platform that allows hosts and guest to connect it seems not to be a requirement form teh platform to request the licence number as 89.8% of the listings don´t count with a registered licence.*

**FIX** on the price I will calculate the average price and edit the missing figures. Otherwise the EDA will present outliers on 0 values, which  is not realistic.

**FIX** on the licence number they will remain as they are. It is already an interesting insight as it is. 

**FIX** df_listing.st had 'id' as 'listing_id'. Changing column name to match with the rest of the dataset columns

In [28]:
# Renaming the column 'id' to 'listing_id'

df_listing_st.rename(columns={'id': 'listing_id'}, inplace=True)

In [29]:
#Checking changes on df_listing_st

df_listing_st.head()

listing_id                                               name    host_id  \
0       21853                               Bright and airy room      83531   
1       24805                             Gran Via Studio Madrid  346366726   
2       40718  COZY APARTMENT CIBELES-GRAN VIA-PTA DEL SOL-CH...   34217435   
3       40916                    Holiday Apartment Madrid Center     130907   
4       30320                        Great Vacational Apartments     130907   

  host_name neighbourhood_group neighbourhood  latitude  longitude  \
0     Abdel              Latina      Cármenes  40.40381   -3.74130   
1    Alvaro              Centro   Universidad  40.42183   -3.70529   
2      Luis              Centro      Justicia  40.41969   -3.69736   
3      Dana              Centro   Universidad  40.42247   -3.70577   
4      Dana              Centro           Sol  40.41476   -3.70418   

         room_type  price  minimum_nights  number_of_reviews last_review  \
0     Private room   31.0               4                 33  2018-07-15   
1  Entire home/apt  104.0               5                 43  2024-06-01   
2  Entire home/apt  114.0               2                150  2024-05-30   
3  Entire home/apt    NaN               5                 49  2021-12-11   
4  Entire home/apt    NaN               5                172  2022-09-26   

   reviews_per_month  calculated_host_listings_count  availability_365  \
0               0.28                               2                83   
1               0.37                               1               113   
2               0.89                              15               310   
3               0.30                               3                 0   
4               1.01                               3                 0   

   number_of_reviews_ltm license  
0                      0     NaN  
1                     13     NaN  
2                     20     NaN  
3                      0     NaN  
4                      0     NaN

In [33]:
# Renaming the column 'id' to 'listing_id'

df_listings.rename(columns={'id': 'listing_id'}, inplace=True)

In [34]:
#Checking changes on df_listings

df_listings.head()

listing_id                                       listing_url  \
0  1104967621421098322  https://www.airbnb.com/rooms/1104967621421098322   
1   930941073152690361   https://www.airbnb.com/rooms/930941073152690361   
2             51647382             https://www.airbnb.com/rooms/51647382   
3  1026072412626070839  https://www.airbnb.com/rooms/1026072412626070839   
4   766423030884644417   https://www.airbnb.com/rooms/766423030884644417   

        scrape_id last_scraped       source                              name  \
0  20240615011404   2024-06-15  city scrape  Habitación de lujo, baño privado   
1  20240615011404   2024-06-15  city scrape            habitación para pareja   
2  20240615011404   2024-06-15  city scrape            Buenos Aires en Madrid   
3  20240615011404   2024-06-15  city scrape            Habitación Madrid Rio!   
4  20240615011404   2024-06-15  city scrape           habitación para turista   

                                         description  neighborhood_overview  \
0                                                NaN                    NaN   
1  Rent room for loose nights or weekends  to a c...                    NaN   
2  This special place is close to everything, mak...  The neighborhood is m   
3  Enjoy the simplicity of this quiet and central...                    NaN   
4  Enjoy a stylish experience at this centrally-l...                    NaN   

                                         picture_url    host_id  ...  \
0  https://a0.muscache.com/pictures/miso/Hosting-...  565307927  ...   
1  https://a0.muscache.com/pictures/85f0b383-e8f8...  442944608  ...   
2  https://a0.muscache.com/pictures/db018af4-f622...  252336445  ...   
3  https://a0.muscache.com/pictures/miso/Hosting-...  545435841  ...   
4  https://a0.muscache.com/pictures/miso/Hosting-...    3882456  ...   

  review_scores_communication review_scores_location review_scores_value  \
0                        5.00                   5.00                5.00   
1                        5.00                   4.67                5.00   
2                        4.83                   4.51                4.61   
3                        5.00                   5.00                5.00   
4                         NaN                    NaN                 NaN   

  license instant_bookable calculated_host_listings_count  \
0     NaN                f                              1   
1     NaN                f                              2   
2     NaN                t                              2   
3     NaN                f                              1   
4     NaN                f                              1   

  calculated_host_listings_count_entire_homes  \
0                                           0   
1                                           0   
2                                           0   
3                                           0   
4                                           0   

  calculated_host_listings_count_private_rooms  \
0                                            1   
1                                            2   
2                                            2   
3                                            1   
4                                            1   

  calculated_host_listings_count_shared_rooms reviews_per_month  
0                                           0              0.61  
1                                           0              0.28  
2                                           0              5.40  
3                                           0              0.14  
4                                           0               NaN  

[5 rows x 75 columns]

In [35]:
# MERGING DF_LISTINGS AND DF_LISTINGS_ST

# Select specific columns from df_listings
df_listings_subset = df_listings[['listing_id', 'listing_url', 'description', 'host_is_superhost', 
                                  'host_total_listings_count', 'host_verifications', 'amenities', 
                                  'beds', 'accommodates', 'bathrooms_text']]

In [37]:
# Merge the two DataFrames on 'listing_id'

df_listings_merged = pd.merge(df_listings_subset, df_listing_st, on='listing_id', how='inner')

In [38]:
#Checking data on new df_listings_merged

df_listings_merged.head()

listing_id                                       listing_url  \
0  1104967621421098322  https://www.airbnb.com/rooms/1104967621421098322   
1   930941073152690361   https://www.airbnb.com/rooms/930941073152690361   
2             51647382             https://www.airbnb.com/rooms/51647382   
3  1026072412626070839  https://www.airbnb.com/rooms/1026072412626070839   
4   766423030884644417   https://www.airbnb.com/rooms/766423030884644417   

                                         description host_is_superhost  \
0                                                NaN                 f   
1  Rent room for loose nights or weekends  to a c...                 f   
2  This special place is close to everything, mak...                 f   
3  Enjoy the simplicity of this quiet and central...                 f   
4  Enjoy a stylish experience at this centrally-l...                 f   

   host_total_listings_count                host_verifications  \
0                        1.0                ['email', 'phone']   
1                        4.0                         ['phone']   
2                        2.0                ['email', 'phone']   
3                        1.0                ['email', 'phone']   
4                        4.0  ['email', 'phone', 'work_email']   

                                           amenities  beds  accommodates  \
0                                                 []   0.0             2   
1                                           ["Wifi"]   1.0             2   
2  ["Outdoor dining area", "TV", "Free parking on...   3.0             1   
3        ["Lock on bedroom door", "Wifi", "Kitchen"]   2.0             3   
4  ["Washer", "Lock on bedroom door", "Air condit...   1.0             1   

  bathrooms_text  ...     room_type  price minimum_nights number_of_reviews  \
0         1 bath  ...  Private room   68.0              1                 2   
1         1 bath  ...  Private room   16.0              1                 3   
2  1 shared bath  ...  Private room   25.0              1               186   
3  1 shared bath  ...  Private room   26.0              2                 1   
4  1 shared bath  ...  Private room  100.0              7                 0   

  last_review  reviews_per_month  calculated_host_listings_count  \
0  2024-05-08               0.61                               1   
1  2023-10-04               0.28                               2   
2  2024-05-29               5.40                               2   
3  2023-11-20               0.14                               1   
4         NaN                NaN                               1   

  availability_365  number_of_reviews_ltm  license  
0                2                      2      NaN  
1              185                      3      NaN  
2              356                     65      NaN  
3               12                      1      NaN  
4              364                      0      NaN  

[5 rows x 27 columns]

In [43]:
# from df_listings_merged calculate the median (50th percentile) of the 'price' column, ignoring NaN values

median_price = df_listings_merged['price'].median()
df_listings_merged['price'].fillna(median_price, inplace=True)

C:\Users\carol\AppData\Local\Temp\ipykernel_37576\2006683898.py:4: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_listings_merged['price'].fillna(median_price, inplace=True)


In [44]:
# Checking prices correction

missing_values=df_listings_merged.isnull().sum()
print("Missing Values in Each Column:\n", missing_values)


Missing Values in Each Column:
 listing_id                            0
listing_url                           0
description                         966
host_is_superhost                   155
host_total_listings_count             4
host_verifications                    4
amenities                             0
beds                               5359
accommodates                          0
bathrooms_text                       36
name                                  0
host_id                               0
host_name                             4
neighbourhood_group                   0
neighbourhood                         0
latitude                              0
longitude                             0
room_type                             0
price                                 0
minimum_nights                        0
number_of_reviews                     0
last_review                        5856
reviews_per_month                  5856
calculated_host_listings_count        0
availabi

In [24]:
# Checking for any missing the df_reviews

missing_values=df_reviews.isnull().sum()
print("Missing Values in Each Column:\n", missing_values)

Missing Values in Each Column:
 listing_id        0
id                0
date              0
reviewer_id       0
reviewer_name     2
comments         93
dtype: int64


In [25]:
# Checking for duplicates df_reviews

duplicate_rows = df_reviews.duplicated().sum()
print("Number of Duplicate Rows:", duplicate_rows)

Number of Duplicate Rows: 0


**NOTE** There were 93 reviews missing a comments and no duplicates. No further action taken 

# 4.  UNDERSTAND YOUR DATA

Develop a basic understanding of your data set by reviewing the variables and performing basic descriptive statistical analysis

In [46]:
#NEW EDA with corrected priced from the merged DF

df_listings_merged_EDA = df_listing_st.drop(columns=['listing_id', 'host_id']).describe()


In [48]:
df_listings_merged_EDA.describe()

latitude     longitude         price  minimum_nights  \
count      8.000000      8.000000      8.000000        8.000000   
mean    3395.821695   3362.732586   5422.056394     3511.915832   
std     9506.830185   9520.189782   9808.013698     9467.991207   
min        0.023563     -3.833071      8.000000        1.000000   
25%       40.389854     -3.702697     90.000000        1.750000   
50%       40.420959     -3.689759    144.676966        5.798748   
75%       40.457742     -2.652477   5474.322917      304.296867   
max    26924.000000  26924.000000  21620.000000    26924.000000   

       number_of_reviews  reviews_per_month  calculated_host_listings_count  \
count           8.000000           8.000000                        8.000000   
mean         3521.342891        2638.050028                     3415.590644   
std          9462.994605        7446.830249                     9499.328472   
min             0.000000           0.010000                        1.000000   
25%             7.750000           1.047500                        2.500000   
50%            46.099057           1.925113                       21.791376   
75%           327.658761           9.117500                      119.106800   
max         26924.000000       21068.000000                    26924.000000   

       availability_365  number_of_reviews_ltm  
count          8.000000               8.000000  
mean        3503.951619            3406.356038  
std         9463.858809            9502.996203  
min            0.000000               0.000000  
25%          103.701857               3.000000  
50%          155.005237              16.899699  
75%          294.500000              82.786680  
max        26924.000000           26924.000000

**NOTE** 50% of the prices are around  144EUR a night. 
Lat and longitude shows that Percentiles shows that most accomodations are located very close to one another. 
Median minimum number of booked nights are 5. with a 75% percentile of 304 days (long term bookings)
Host hava a median of 21 listings. Meaning most booking are done through agencies rather than individuals owners. 

In [50]:
# Checking data types

df_listings_merged.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 26924 entries, 0 to 26923
Data columns (total 27 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   listing_id                      26924 non-null  int64  
 1   listing_url                     26924 non-null  object 
 2   description                     25958 non-null  object 
 3   host_is_superhost               26769 non-null  object 
 4   host_total_listings_count       26920 non-null  float64
 5   host_verifications              26920 non-null  object 
 6   amenities                       26924 non-null  object 
 7   beds                            21565 non-null  float64
 8   accommodates                    26924 non-null  int64  
 9   bathrooms_text                  26888 non-null  object 
 10  name                            26924 non-null  object 
 11  host_id                         26924 non-null  int64  
 12  host_name                       

In [51]:
#Checking for mix types 

# Identify columns with object data types
object_columns = df_listings_merged.select_dtypes(include=['object']).columns
print("Columns with Object Data Type:\n", object_columns)

Columns with Object Data Type:
 Index(['listing_url', 'description', 'host_is_superhost', 'host_verifications',
       'amenities', 'bathrooms_text', 'name', 'host_name',
       'neighbourhood_group', 'neighbourhood', 'room_type', 'last_review',
       'license'],
      dtype='object')


In [56]:
# List of columns to convert to string
columns_to_convert = [
    'listing_url', 'description', 'host_is_superhost', 'host_verifications',
    'amenities', 'bathrooms_text', 'name', 'host_name',
    'neighbourhood_group', 'neighbourhood', 'room_type', 'last_review',
    'license']

In [57]:
# Convert the specified columns to string
df_listings_merged[columns_to_convert] = df_listings_merged[columns_to_convert].astype(str)

# Display the data types to confirm the conversion
print(df_listings_merged.dtypes)

listing_id                          int64
listing_url                        object
description                        object
host_is_superhost                  object
host_total_listings_count         float64
host_verifications                 object
amenities                          object
beds                              float64
accommodates                        int64
bathrooms_text                     object
name                               object
host_id                             int64
host_name                          object
neighbourhood_group                object
neighbourhood                      object
latitude                          float64
longitude                         float64
room_type                          object
price                             float64
minimum_nights                      int64
number_of_reviews                   int64
last_review                        object
reviews_per_month                 float64
calculated_host_listings_count    

In [58]:
# EXPORT df_li

df_listings_merged.to_pickle(os.path.join(path, 'Prepared data', 'df_listings_merged.pkl'))